### ISSUE TITLE
embark-collect: Uniform default action behavior

### ISSUE THREAD

**Issue by minad** [oantolin/embark#494]
I just saw your comment https://old.reddit.com/r/emacs/comments/u9lb0s/followpreview_in_embarkcollect_buffer/i5usk05/ and it made me wonder why we are not using a uniform behavior for the `embark-collect` buffers.

We could always use a predefined default action for `RET`. When collecting candidates, the default action is forgotten. Collection is truly only about the candidates. The default action is usually always about viewing the candidates as asked in the reddit question.
This shouldn't be a problem since the other actions like `delete-file` are available anyway as action. I believe that this change makes the collect functionality slightly simpler and potentially easier to understand and also more like Dired or Ibuffer. Maybe the current design was only a necessity for the use case of Embark as a completion system, which is not that relevant anymore?

Currently the collect buffers contain the default action in their buffer name, e.g., `*Embark Collect: M-x find-file RET ~/*`. Instead the buffer name should only contain the type of the candidates, e.g., `*Embark Collect: Files*`. Note that there exists some other awkwardness currently. If you run collect in a collect buffer recursively you end up with a buffer `*Embark Collect: *Embark Collect: *Embark Collect: M-x find-file RET ~/***`.

Note that there are a few target types where we don't specify a default action, e.g., the `embark-package-map`. These keymaps would require fixing. We could even throw an error at compile time in the `embark-define-keymap` if neither `:parent` nor a `RET` action is specified.

**Comment by minad** [comment]
I thought about this a bit more. Usually I use collect buffers if I want some candidates to stay around for longer. In those cases I almost exclusively want to perform just a viewing operation on the candidates. For this use case, which as I claim, is the main use case one could very well just use a predefined default action. However such a workflow is usually initiated from a find/search operation anway such that the difference wouldn't be noticeable. Anyway the emphasis here is on having candidates staying around for longer, such that the initiating command is less and less relevant. It is only about the candidates then.

Now there are also other workflows, e.g., starting collect from `delete-file` to delete multiple files afterwards by pressing RET. As far as I recall, I've never used such a workflow. I usually use the Dired export also because of the automatic update of the candidate list. If I would use collect instead, I would now take advantage of the relatively recent `embark-collect-mark` functionality and `embark-act-all` to execute the destructive operation in a more controlled manner. 

Given that we have `embark-act-all` and marking and that we've gotten rid of `embark-collect-completions`, it seems to me that the default action override of `embark-collect` can now be considered an obsolete artifact.

**Comment by oantolin** [comment]
Before I address the main point, let me comment on the buffer names:

> Currently the collect buffers contain the default action in their buffer name

Technically not quite: if there is an entry in `embark-default-action-overrides` then the command in the buffer name may not be the default action. Indeed, I wasn't thinking of the default action when I decided to give the buffers that kind of name: I was thinking of revert! The name tells you what command will be run if you revert the buffer.

>  If you run collect in a collect buffer recursively you end up with a buffer `*Embark Collect: *Embark Collect: *Embark Collect: M-x find-file RET ~/***`.

That name does look funny, but it accurately informs you that nothing interesting would happen if you revert that buffer.

In summary, the buffer name isn't meant to specify the default action but to tell you how you got the buffer, which also tells you what will happen upon reverting.

**Comment by oantolin** [comment]
Now, for your main point that maybe we should not use the "origin command" as the default action:

The current design lets a user use the approach you suggested already! If someone does not want to risk a dangerous operation being the default action, all they need to do is set a default action override. If you set the default action for `file` to be `find-file`, boom, no more "RET deleted my file!" problems.

> Now there are also other workflows, e.g., starting collect from delete-file to delete multiple files afterwards by pressing RET. As far as I recall, I've never used such a workflow.

Using a "non-viewing" default action is a rare occurrence for me too. I think the only I do use with some regularity is to attach files to an email. I quite often need to attach more than one file to an email, and in that case they are usually all in the same directory. I've used all of these methods to attach multiple files:

- Use `embark-dwim` on each without closing the minibuffer.
- If I can get exactly the ones I want to match, then `<embark-act-all> RET`.
- If I notice I'm not quite sure which files I need to attach (often because they have non-descriptive names, like random strings of numbers, which sometimes happens with pictures or with downloaded academic papers in PDF), I'll use `embark-collect`. In that buffer I use `<embark-act> f` to view the file and `RET` to attach it.

If it were not for the fact that `mml-attach-file` becomes the default action, I wouldn't have a convenient way of attaching those files in the last method. (I know there is a `gnus-dired-attach` command, I could use to mark and attach the files in a dired buffer, but I often forget it exists, I don't have it already autoloaded, and I haven't memorized it's key binding.)

I've also used collect buffers where the default action is `package-install` or `rename-file` (you get prompted each time for the new name, so I only use this if I'm sending files to a different diectories). But I do this very rarely: I almost always produce collect buffers from the "viewing command" for the type.

**Comment by oantolin** [comment]
If we did adopt what you call the uniform approach, maybe it should only be done in collect buffers. I quite like using `embark-dwim` in the minibuffer to mean "pretend I pressed RET but don't close the completion session!".

Maybe that would work? Have the default action override only apply to collect buffers? We could rename the user option to better reflect its purpose if we do that.

**Comment by oantolin** [comment]
Another possibility for the hypothetical follow mode from that reddit comment is to refuse to activate or at least warn the user if you try to activate it in a collect buffer where the default action is not given by `embark-default-action-overrides`.

**Comment by minad** [comment]
1. I agree with your reasoning about buffer naming. It is good to know where the candidates are coming from. However I wonder why the name is so verbose. Why do we need the `M-x` and the `RET`? Wouldn't it suffice to call the buffer `*Embark Collect: find-file*`?
2. Regarding the default action override - I don't think this can be used universally. First it will also override the default action in the minibuffer where I definitely would like to keep the current behavior. Furthermore I have no way of saying "Always use the action RET in the collect buffers" for all types. As I understood the default action overrides until now, they are meant to fix broken actions, which wouldn't work otherwise due to the special minibuffer injection requirements. For example embark-consult-goto-location is significantly more efficient.
3. Your attach-file, package-install and rename-file workflows are plausible but I think these could be easily fixed by binding the appropriate actions to the action map (if they are not already there). It may not be as convenient as pressing RET but at least the behavior will be uniform. Furthermore as long as you are in the minibuffer and started with the respective command RET will still stay functional.
4. To be clear - I argue only about collect buffers. In the minibuffer everything should stay as is with the `embark-dwim`. But in the collect buffer I consider the current behavior problematic since it deviates from Dired/Ibuffer and it is also potentially dangerous. If we change this to be uniform then collect becomes also easier to explain - it is just like a generalized Dired, nothing special.
5. I wouldn't add a follow mode to be honest or at least not add special casing if such a mode is added. If we change the collect buffer to use uniform default actions, we can add a follow mode and it wouldn't have serious dangerous downsides. But the problem of preview still remains if you don't want to end up with many open buffers. Maybe the preview base functionality should be factored out of Consult (see also https://github.com/minad/consult/issues/439), since getting this right is not really trivial. There are many things to consider when previewing files to ensure that the UI stays responsible and to ensure that no expensive modes are started.

**Comment by minad** [comment]
@oantolin Did you think about this? I looked into the code and both revert and the default action rely on the buffer-local variable `embark--command`. In order to make the RET behavior more uniform one would have to ignore the command in `embark--default-action` if triggered from the collect buffer. This would certainly complicate things or require some refactoring such that it still feels natural.

**Comment by oantolin** [comment]
I'm not convinced yet. I think it would introduce an easily-forgotten discontinuity. I can imagine needing to run some operation on some item, so I fire up `M-x operation` and try to narrow the match to the item I want. But maybe I notice then there are a couple of items more I also want to operate on. No problem, I'll just use `embark-act-all RET`. But maybe matching exactly the ones I want is a little tricky, so I decide to use `embark-collect` and mark the correct items. I'd still want `RET` to refer to `M-x operation` at that last stage.

Also, in `embark-live` buffers, it would be weird if `RET` didn't mean the originating command: the minibuffer is still open for part of the life of the embark live buffer! Should the meaning of RET always ignore the minibuffer? Should RET run the originating command while the minibuffer is open and switching to something else once you quit?

**Comment by oantolin** [comment]
One problem is that it is a little hard to find out what the default action is without actually running it!

**Comment by minad** [comment]
I agree that `embark-live` poses a problem in the case of the minibuffer, but in the more realistic scenario of live TOCs for normal buffers, the RET is always the default action as specified in the keymap since there hasn't been a minibuffer. I think it boils down to how you use embark collect. I often use it to collect candidates and keep them around for longer and then I almost always want the default action to just display the candidate (the "uniform" behavior). I understand your point and I am also somewhat torn. One thing I find important is also the symmetry with Dired and Ibuffer, where RET also displays the candidates.

**Comment by oantolin** [comment]
But it's not just `embark-live`. I find the first scenario I mentioned, in which `embark-collect` is used for the ability to mark candidates, more troublesome.

**Comment by oantolin** [comment]
If you follow the "litte thing-manager" strategy of using display-type commands as the entry point to a type of candidates, and from there using other commands as actions, or exporting or collecting. Then there is no difference anyway: your collect buffers will have the display-type command as the default command.

I think I do that a lot. I don't even have global key bindings for things like `delete-file` or `package-delete`, I always run them as actions: `delete-file` normally from `find-file` but occasionally from `mml-attach-file`; and `package-delete` from `describe-package`.

**Comment by minad** [comment]
> If you follow the "litte thing-manager" strategy of using display-type commands as the entry point to a type of candidates, and from there using other commands as actions, or exporting or collecting. Then there is no difference anyway: your collect buffers will have the display-type command as the default command.

Yes, that's right and the proposal here is to just bless these entry points (display-type commands) to be always used in collect buffers, such that the collect buffer works like Dired.

> I think I do that a lot. I don't even have global key bindings for things like delete-file or package-delete, I always run them as actions: delete-file normally from find-file but occasionally from mml-attach-file; and package-delete from describe-package.

Yes, I use the same workflow.

**Comment by oantolin** [comment]
> \1. `rename-file`

I see. I'd start with `find-file`, not `rename-file`, just out of laziness: I have a key binding for `find-file` but not for `rename-file`.This issue does matter more if you run commands like `rename-file` directly.

**Comment by minad** [comment]
> One thing I like about the marking functionality in collect buffers is that then you don't really need it in the completion UIs. But I feel like marking loses some of its power under the uniform default action proposal.

That's right and I agree that this doesn't sound so good, but maybe it is also an edge case? I think the point here is that you use the collect buffer like a completion context. Then you want RET to behave like the original command.

On the other hand if you look at it from the other perspective where a candidate collection is decoupled from its original command, then a uniform RET feels more natural. I think in most buffers RET either means insert a newline or it means open/display something at point (in particular in special mode buffers or links at point in org).

**Comment by minad** [comment]
It also seems to matter where one is coming from. For me Embark was mostly an action system since I never used it for completion. As such the completion-style RET behavior in collect buffers never mattered to me (while it does of course when acting in the minibuffer). In contrast, for you this mattered (or matters) much more. I mentioned this before, where I asked if the current behavior could be considered a remnant of the usage of Embark as completion UI. Overall I am torn which behavior is better, since as we have it now it is also natural and consistent. I just wanted to put this up for discussion. There is no reason to do anything, no urgency. But it should probably be settled before 1.0 in either way ;)

**Comment by oantolin** [comment]
> One could however argue that maybe the mode line of such special buffers should blink in neo green with red bold text such that you don't accidentally RET delete :-P

Or, you can keep the default setting of `(alist-get 'delete-file embark-pre-action-hooks) = '(embark--confirm)`. 😉  I guess the confirmation also makes me worry less about keeping things as they are.

Also, since the originating command appears in the buffer name, it's at least not that big of a surprise when RET runs it.

If we do change things I'd propose just changing the priority of the possible sources of the default action in collect buffers. Currently they are:

1. `embark-default-action-overrides`
2. the originating command, `embark--comand`.
3. the binding of RET in the keymap registered for the type in `embark-keymap-alist`.

We could just swap 2 and 3 in collect buffers. What I like about having `embark--command` as a fallback is that I want RET to do *something*, and in case a completion table does not specify a category, or even if it does, but there is no embark keymap for it, I still want something plausibly useful to happen when I press RET.

I think I'd go for the swap, try it a while, and if I still yearn for the current behavior (which seems unlikely: I hardly take advantage of the difference), then maybe I could add a user option to choose between the two behaviors.

**Comment by minad** [comment]
Okay, the swap is essentially what I proposed. But then we lose the advantage of creating the "special collect buffers" as I mentioned, which make the original command accessible via RET. One could also bind the original command always to M-RET and the default to RET and swap the behavior in the minibuffer to make the confusion complete.

**Comment by oantolin** [comment]
To be honest, I don't find the current behavior problematic at all. I'd be marginally in favor of keeping it. You can see in the buffer name what the default action is (if not overridden...). In fact, maybe it would be confusing if RET does *not* run the command in the buffer name! 😛

But I also don't oppose the change, if you think it's important I'd definitely try it.

**Comment by oantolin** [comment]
> Okay, the swap is essentially what I proposed. But then we lose the advantage of creating the "special collect buffers" as I mentioned, which make the original command accessible via RET. One could also bind the original command always to M-RET and the default to RET and swap the behavior in the minibuffer to make the confusion complete.

I did think of that (even down to picking the same M-RET binding!), but that's probably too much.

**Comment by minad** [comment]
Well maybe it is worth to try it but after this discussion I also wouldn't mind if the current behavior is kept as is.

**Comment by oantolin** [comment]
> One could however argue that maybe the mode line of such special buffers should blink in neo green with red bold text such that you don't accidentally RET delete :-P

We could color the modeline differently (not *that* differently) for these special buffers. But I'd argue that the buffer names are already enough indication of the special nature. That and having `embark--confirm` for the most dangerous commands is probably enough. For example `rename-file` isn't dangerous at all since you get prompted for a destination file name and can cancel at that point.

**Comment by hmelman** [comment]
> > \1. `rename-file`
> 
> I see. I'd start with `find-file`, not `rename-file`, just out of laziness: I have a key binding for `find-file` but not for `rename-file`. This issue does matter more if you run commands like `rename-file` directly.

FWIW my typical rename workflow is `C-x C-j R`, that's `dired-jump` and then `R` in `dired` for rename.  Type the name and then `C-x C-k` to go back to what I was doing. :)

I haven't followed this too closely.  I'm all in favor of the collect buffer mirring `dired` whenever possible (e.g., marking).

> On the other hand if you look at it from the other perspective where a candidate collection is decoupled from its original command, then a uniform RET feels more natural. 

I think this sounds more like what I'd want.  It seems to me I might collect a buffer with one command (say search like `consult-locate`) and then want to do stuff with the results, either further refine or act on.  I also tend to keep collect buffers around a little.  Not a long time like an outline table of contents, but I might do a search and then look at individual files or something in a manual and move back to the results.  So the originating command is less useful to me.  Lastly I probably don't trust knowing what default action is on RET so I'd probably stick to other more explicit bindings to do things :)

Is there a recipe I could use with the current melpa release to play with the kind of workflows you're describing?  I haven't really played with `embark-act-all` yet.

**Comment by oantolin** [comment]
Having made another impassioned defense of user over at #495, I think I'm a little more inclined than before to keep the current behavior, since in your terminology, @minad, it allows for both uniform and special collect buffers.

But it might be nice to add a way to mark a default action override as only applying to collect buffers.

@hmelman The collect buffer header line is currently off default but if you turn it on it is the one provided by tabulated list mode. It says "Candidate" in one column, "Annotation" in the other and let's you sort by either by clicking on a little triangle.
